File: app/src/main/java/tpcreative/co/qrscanner/common/view/CustomViewPager.java
package tpcreative.co.qrscanner.common.view;

import android.content.Context;
import android.support.v4.view.MotionEventCompat;
import android.support.v4.view.ViewPager;
import android.util.AttributeSet;
import android.view.MotionEvent;

public class CustomViewPager extends ViewPager {
    float mStartDragX;
    OnSwipeOutListener mOnSwipeOutListener;

    public CustomViewPager(Context context) {
        super(context);
    }

    public CustomViewPager(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public void setOnSwipeOutListener(OnSwipeOutListener listener) {
        mOnSwipeOutListener = listener;
    }

    private void onSwipeOutAtStart() {
        if (mOnSwipeOutListener!=null) {
            mOnSwipeOutListener.onSwipeOutAtStart();
        }
    }

    private void onSwipeMove(){
        if (mOnSwipeOutListener!=null) {
            mOnSwipeOutListener.onSwipeMove();
        }
    }

    private void onSwipeOutAtEnd() {
        if (mOnSwipeOutListener!=null) {
            mOnSwipeOutListener.onSwipeOutAtEnd();
        }
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        switch(ev.getAction() & MotionEventCompat.ACTION_MASK){
            case MotionEvent.ACTION_DOWN:
                mStartDragX = ev.getX();
                break;
        }
        return super.onInterceptTouchEvent(ev);
    }

    @Override
    public boolean onTouchEvent(MotionEvent ev){
        if(getCurrentItem()==0 || getCurrentItem()==getAdapter().getCount()-1){
            final int action = ev.getAction();
            float x = ev.getX();
            switch(action & MotionEventCompat.ACTION_MASK){
                case MotionEvent.ACTION_MOVE:
                     onSwipeMove();
                    break;
                case MotionEvent.ACTION_UP:
                    if (getCurrentItem()==0 && x>mStartDragX) {
                        onSwipeOutAtStart();
                    }
                    if (getCurrentItem()==getAdapter().getCount()-1 && x<mStartDragX){
                        onSwipeOutAtEnd();
                    }
                    break;
            }
        }else{
            mStartDragX=0;
        }
        return super.onTouchEvent(ev);
    }

    public interface OnSwipeOutListener {
        void onSwipeOutAtStart();
        void onSwipeOutAtEnd();
        void onSwipeMove();
    }
}

File: app/src/main/java/tpcreative/co/qrscanner/common/view/CustomViewPager.kt
package tpcreative.co.qrscanner.common.view

import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import androidx.core.view.MotionEventCompat
import androidx.viewpager.widget.ViewPager

class CustomViewPager : ViewPager {
    var mStartDragX = 0f
    var mOnSwipeOutListener: OnSwipeOutListener? = null
    private var mDisable = false
    constructor(context: Context) : super(context) {}
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {}

    fun setOnSwipeOutListener(listener: OnSwipeOutListener?) {
        mOnSwipeOutListener = listener
    }

    private fun onSwipeOutAtStart() {
        if (mOnSwipeOutListener != null) {
            mOnSwipeOutListener?.onSwipeOutAtStart()
        }
    }

    private fun onSwipeMove() {
        if (mOnSwipeOutListener != null) {
            mOnSwipeOutListener?.onSwipeMove()
        }
    }

    private fun onSwipeOutAtEnd() {
        if (mOnSwipeOutListener != null) {
            mOnSwipeOutListener?.onSwipeOutAtEnd()
        }
    }

    fun setSwipeableDisable(swipeableDisable: Boolean) {
        this.mDisable = swipeableDisable
    }

    override fun onInterceptTouchEvent(ev: MotionEvent?): Boolean {
        if (this.mDisable) {
            return false
        }
        when (ev?.action?.and(MotionEventCompat.ACTION_MASK)) {
            MotionEvent.ACTION_DOWN -> mStartDragX = ev.x
        }
        return super.onInterceptTouchEvent(ev)
    }

    override fun onTouchEvent(ev: MotionEvent?): Boolean {
        if (this.mDisable) {
            return false
        }
        if (currentItem == 0 || currentItem == (adapter?.count ?:0) - 1) {
            val action = ev?.action
            val x = ev?.x
            if (action != null) {
                when (action and MotionEventCompat.ACTION_MASK) {
                    MotionEvent.ACTION_MOVE -> onSwipeMove()
                    MotionEvent.ACTION_UP -> {
                        if (currentItem == 0 && (x ?:0F) > mStartDragX) {
                            onSwipeOutAtStart()
                        }
                        if (currentItem == (adapter?.count ?:0) - 1 && (x ?:0F) < mStartDragX) {
                            onSwipeOutAtEnd()
                        }
                    }
                }
            }
        } else {
            mStartDragX = 0f
        }
        return super.onTouchEvent(ev)
    }

    interface OnSwipeOutListener {
        fun onSwipeOutAtStart()
        fun onSwipeOutAtEnd()
        fun onSwipeMove()
    }
}